In [1]:
# Add the Pandas dependency.
import pandas as pd

# Load the CSV File (My Uber Drives - 2016.csv)

In [23]:
# Files to load
uber_data_to_load = "Resources/My Uber Drives - 2016.csv"

# Read the Uber Drives Data File

In [4]:
uber_data_df = pd.read_csv(uber_data_to_load)
uber_data_df

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*         PURPOSE*  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1156 rows x 7 columns]

In [30]:
print(uber_data_df)

           START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*         PURPOSE*  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0  

In [25]:
len(uber_data_df)

1156

In [24]:
uber_data_df.dtypes

START_DATE*     object
END_DATE*       object
CATEGORY*       object
START*          object
STOP*           object
MILES*         float64
PURPOSE*        object
dtype: object

In [5]:
# Determine if there are any missing values in the uber drives data.
uber_data_df.count()

START_DATE*    1156
END_DATE*      1155
CATEGORY*      1155
START*         1155
STOP*          1155
MILES*         1156
PURPOSE*        653
dtype: int64

In [6]:
# Determine if there are any missing values in the uber drives data.
uber_data_df.isnull()

START_DATE*  END_DATE*  CATEGORY*  START*  STOP*  MILES*  PURPOSE*
0           False      False      False   False  False   False     False
1           False      False      False   False  False   False      True
2           False      False      False   False  False   False     False
3           False      False      False   False  False   False     False
4           False      False      False   False  False   False     False
...           ...        ...        ...     ...    ...     ...       ...
1151        False      False      False   False  False   False     False
1152        False      False      False   False  False   False     False
1153        False      False      False   False  False   False     False
1154        False      False      False   False  False   False     False
1155        False       True       True    True   True   False      True

[1156 rows x 7 columns]

### Finding:

Row 1155 indicates null values for the following columns: "END DATE*", "CATEGORY*", "START*" and "STOP*".

# (But why?--I don't know at this time.)

In [7]:
# Determine if there are any missing values in the uber drives data.
uber_data_df.isnull().sum()

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

### Finding:

The "PURPOSE*" column has a total 503 missing values.

In [8]:
# Determine if there are not any missing values in the uber drives data.
uber_data_df.notnull()

START_DATE*  END_DATE*  CATEGORY*  START*  STOP*  MILES*  PURPOSE*
0            True       True       True    True   True    True      True
1            True       True       True    True   True    True     False
2            True       True       True    True   True    True      True
3            True       True       True    True   True    True      True
4            True       True       True    True   True    True      True
...           ...        ...        ...     ...    ...     ...       ...
1151         True       True       True    True   True    True      True
1152         True       True       True    True   True    True      True
1153         True       True       True    True   True    True      True
1154         True       True       True    True   True    True      True
1155         True      False      False   False  False    True     False

[1156 rows x 7 columns]

In [9]:
# Determine if there are not any missing values in the uber drives data.
uber_data_df.notnull().sum()

START_DATE*    1156
END_DATE*      1155
CATEGORY*      1155
START*         1155
STOP*          1155
MILES*         1156
PURPOSE*        653
dtype: int64

# Handle Missing Data [use dropna() method] (Reference: Module 4.5.2)

https://courses.bootcampspot.com/courses/1225/pages/4-dot-5-2-handle-missing-data?module_item_id=497375

Column G--PURPOSE--in the (My Uber Drives - 2016.csv) file has missing data/no entries (i.e., null values).
Drop these entries as well as the entries from above:
(Row 1155 indicates null values for the following columns: "END DATE*", "CATEGORY*", "START*" and "STOP*".)

Pandas documentation on the dropna() method: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html

In [16]:
uber_data_drop_purpose_df = uber_data_df.dropna()
uber_data_drop_purpose_df

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                 STOP*  MILES*         PURPOSE*  
0          Fort Pierce     5.1   Meal/Entertain  
2          Fort Pierce     4.8  Errand/Supplies  
3          Fort Pierce     4.7          Meeting  
4      West Palm Beach    63.7   Customer Visit  
5      West Palm Beach     4.3   Meal/Entertain  
...                ...     ...              ...  
1150           Kar?chi     0.7          Meeting  
1151  Unknown Location     3.9   Temporary Site  
1152  Unknown Location    16.2          Meeting  
1153           Gampaha     6.4   Temporary Site  
1154         Ilukwatta    48.2   Temporary Site  

[653 rows x 7 columns]

# Double check for missing values in the new DataFrame

In [17]:
uber_data_drop_purpose_df.count()

START_DATE*    653
END_DATE*      653
CATEGORY*      653
START*         653
STOP*          653
MILES*         653
PURPOSE*       653
dtype: int64

In [18]:
uber_data_drop_purpose_df.isnull()

START_DATE*  END_DATE*  CATEGORY*  START*  STOP*  MILES*  PURPOSE*
0           False      False      False   False  False   False     False
2           False      False      False   False  False   False     False
3           False      False      False   False  False   False     False
4           False      False      False   False  False   False     False
5           False      False      False   False  False   False     False
...           ...        ...        ...     ...    ...     ...       ...
1150        False      False      False   False  False   False     False
1151        False      False      False   False  False   False     False
1152        False      False      False   False  False   False     False
1153        False      False      False   False  False   False     False
1154        False      False      False   False  False   False     False

[653 rows x 7 columns]

In [19]:
uber_data_drop_purpose_df.isnull().sum()

START_DATE*    0
END_DATE*      0
CATEGORY*      0
START*         0
STOP*          0
MILES*         0
PURPOSE*       0
dtype: int64

In [20]:
uber_data_drop_purpose_df.notnull()

START_DATE*  END_DATE*  CATEGORY*  START*  STOP*  MILES*  PURPOSE*
0            True       True       True    True   True    True      True
2            True       True       True    True   True    True      True
3            True       True       True    True   True    True      True
4            True       True       True    True   True    True      True
5            True       True       True    True   True    True      True
...           ...        ...        ...     ...    ...     ...       ...
1150         True       True       True    True   True    True      True
1151         True       True       True    True   True    True      True
1152         True       True       True    True   True    True      True
1153         True       True       True    True   True    True      True
1154         True       True       True    True   True    True      True

[653 rows x 7 columns]

In [21]:
uber_data_drop_purpose_df.notnull().sum()

START_DATE*    653
END_DATE*      653
CATEGORY*      653
START*         653
STOP*          653
MILES*         653
PURPOSE*       653
dtype: int64

In [22]:
uber_data_drop_purpose_df.dtypes

START_DATE*     object
END_DATE*       object
CATEGORY*       object
START*          object
STOP*           object
MILES*         float64
PURPOSE*        object
dtype: object

In [28]:
len(uber_data_drop_purpose_df)

653

In [29]:
print(uber_data_drop_purpose_df)

           START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
5       1/6/2016 17:15    1/6/2016 17:19  Business   West Palm Beach   
...                ...               ...       ...               ...   
1150   12/31/2016 1:07   12/31/2016 1:14  Business           Kar?chi   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   

                 STOP*  MILES*         PURPOSE*  
0          Fort Pierce     5.1   Meal/Entertain  
2          Fort Pierce     4.8  Err

# Export uber_data_drop_purpose_df as a .csv file

In [39]:
# syntax: df.to_csv(r'Path where you want to store the exported CSV file\File Name.csv')
uber_data_drop_purpose_df.to_csv(r'C:\Users\micha\OneDrive\Desktop\Analysis Projects\Final_project\Resources\Clean Uber Drives - 2016.csv')

# Summary and Personal Thoughts:

### Summary:

The (My Uber Drives - 2016.csv) file was assigned to the variable "uber_data_to_load" and then converted into a DataFrame via the read_csv() method.

The DataFrame was initially assigned/named uber_data_df.

When checking for null values, there were a total of 653 for the file's "PURPOSE*" column.

There were also 1 missing value(s) for the following columns on row 1155: "END DATE*", "CATEGORY*", "START*" and "STOP*".

Pandas dropna() method used on uber_data_df (assigned to uber_data_drop_purpose_df)

0 missing values upon performing uber_data_drop_purpose_df.isnull() and uber_data_drop_purpose_df.isnull().sum().

### Personal Thoughts after reviewing the findings:

Question: Do we really need to include the PURPOSE* column for the analysis (since 1/3 of the entries were dropped)?

In other words, can we continue our analysis focusing on the other more complete columns instead of dropping all the rows that have missing values for the PURPOSE* column?